17 add eoa account creation

JohnGuilding opened the pull request:
- createAccount
- generatePrivateKey
- recoverAccount
- sendTransaction
- getBalance

resolves #19

JohnGuilding committed:
Add eoa account
- createAccount
- generatePrivateKey
- recoverAccount
- sendTransaction
- getBalance

JohnGuilding committed:
Export ExternallyOwnedAccount

JohnGuilding committed:
Update EOA create account function signature & add missed create account logic

JohnGuilding committed:
Use consistent account naming convention

JohnGuilding committed:
Rename EoaAccount to ExternallyOwnedAccount

JohnGuilding committed:
Eoa -> EOA

JohnGuilding committed:
EOA -> Eoa

JohnGuilding committed:
Rename Eoa -> ExternallyOwnedAccount

blakecduncan commented on ethdk/src/ExternallyOwnedAccount/ExternallyOwnedAccount.ts:
```diff
@@ -0,0 +1,96 @@
+    return new ExternallyOwnedAccount({
+      address: signer.address,
+      privateKey: signer.privateKey,
+      recoveryPhrase: signer.mnemonic.phrase,
```
Curious to hear your reasoning for naming the seed phrase a `recoveryPhrase`?  I'm not necessarily against it, I just think of a mnemonic and the private key similarly.  I think they are more for generating a wallet.  Not necessarily for recovering a wallet that exists.

JohnGuilding replied:
Good point, I think I was just thinking about recovery and didn't consider wallet generation. My reasoning was that recoveryPhrase then seemed more helpful (less web3 jargon-y) over mnemonic or seedPhrase. But I agree the mnemonic or seedPhrase better encapsulates the recovery and generation aspect. Will update

blakecduncan reviewed (commented):
💯   This looks good!  Thanks for all the tests.

I just had one question about the recoveryPhrase in the eoa class

JohnGuilding committed:
Account creation handles privateKey & seedPhrase param

JohnGuilding commented on ethdk/src/types/brands.ts:
```diff
@@ -0,0 +1,5 @@
+/** Generic Brand type to allow nominal typing */
+export type Brand<T, B> = T & { __brand: B }
```
Decided to use type branding to discern between a private key and seed phrase

sources
- chatgpt
- https://basarat.gitbook.io/typescript/main-1/nominaltyping - this link has some other solutions if anyone is interested
- https://github.com/kourge/ts-brand

blakecduncan replied:
This is cool, I didn't know this was a thing.  I think this is a cool solution but I am torn on this because I think it may be overcomplicating the pk/seed phrase a little.  For example, I'm not a fan of having to cast the value that we are using in the config object like so.

```typescript
const account = await ExternallyOwnedAccount.createAccount({
      privateKeyOrSeedPhrase: privateKey as PrivateKey,
      network,
})
```

I did some thinking on this over the weekend because I was wondering how we could make this simpler.  I was thinking it would probably be best to have different interfaces for the EOA class.  Something like this

```typescript
type BaseConfig = {
  network?: Network
}

type PrivateKeyEoa = {
  privateKey: string
} & BaseConfig

type SeedPhraseEoa = {
  seedPhrase: string
} & BaseConfig

type EoaConfig = BaseConfig | PrivateKeyEoa | SeedPhraseEoa

static async createAccount(
    config: EoaConfig = {},
): Promise<ExternallyOwnedAccount> {
    if (isPrivateKey(config)) {
      console.log('Private Key:', config.privateKey)
    } else if (isSeedPhrase(config)) {
      console.log('Seed Phrase:', config.seedPhrase)
    }
}
```

I don't have a strong opinion, so if you want to stay with the nominal typing thats cool.  I was just trying to think of a way that we don't have to assert the type of a value we pass to the createAccount function or have a long key name.

blakecduncan replied:
Hmm the only undesirable behavior of my above code is that it doesn't give a typescript error if you have both the seedPhrase and the privateKey like so
```typescript
{
      seedPhrase: 'test',
      privateKey: 'asdf',
      network,
}
```
I guess that's because typescript allows excess properties for objects

JohnGuilding replied:
Agree regarding added type brand adding complexity, casting and long variable names. As mentioned offline:

Changes:
- only allow one type at a time (no `privateKeyOrSeedPhrase`, or even a more abstract `key`). So we don't have to assert on the type and it's clear to the user what they are creating the account with, with either `createAccountFromSeedPhrase` or `createAccountFromPrivateKey`
- `createAccount` method at top level of ethdk is only concerned with sensible defaults. So it now only has a private key parameter, and no seed phrase.

This removes the branding, casting to `PrivateKey`, creates clear separation of concerns with the dedicated static methods on the EOA class, while also allowing a sensible default option through `createAccount`.

JohnGuilding replied:
> I guess that's because typescript allows excess properties for objects

The updated code still doesn't provide the behaviour we're describing here. e.g. I didn't get a compile error with the old parameter name of `privateKeyOrSeedPhrase`, and had to manually debug before I realised the issue in the test set up.

Would be good to find a nice solution to this in a future issue. It's not hard to make a typo like `privateKeu` without noticing, so would be a good DX win

JohnGuilding replied:
We could add more type safety on the account config by creating a dedicated type/interface:

```ts
interface CreateAccountConfig {
  privateKey?: string
  network?: Network
}
```

This throws a compile error when typos are being passed in like `privateKeu`. Issue with this approach from testing it out, is it give less helpful type information and documentation when hovering over the methods:

before:
```ts
(method) ExternallyOwnedAccount.createAccount({ privateKey, network, }?: {
    privateKey?: string;
    network?: Network;
}): ExternallyOwnedAccount
```

after:
```ts
(method) ExternallyOwnedAccount.createAccount({ privateKey, network, }?: CreateAccountConfig): ExternallyOwnedAccount
```

blakecduncan replied:
@JohnGuilding Is there a functional difference between the before and after?  I've always thought the before and after are the same.  The `after` may just be a little more readable since the config is broken out into an interface.  It could also be more reusable if that same CreateAccountConfig is used across multiple methods. But I think the type safety should be the same.

JohnGuilding replied:
Just tested again and yeah you're right, I made a mistake when testing this - I forgot to add a variable for the typo I made, which made me interpret the error wrong lol. So yeah think the only difference is readability

blakecduncan commented on ethdk/src/Ethdk.ts:
```diff
@@ -32,5 +34,14 @@ export async function createAccount<T extends keyof AccountTypeMap>({
+
+  if (accountType === 'eoa') {
+    const account = await ExternallyOwnedAccount.createAccount({
+      privateKeyOrSeedPhrase: privateKey as PrivateKey,
```
I'm not a big fan of the name here `privateKeyOrSeedPhrase`. The reason is if we ever had a different way to create an account we would need to change the name.  And we wouldn't keep adding to the key.  ex: `privateKeyOrSeedPhraseOrSomething`.  If we are going to have a key that can be multiple types I would come up with a more generic name.

I think for this case it isn't a big deal because an eoa account can really only be created with a pk or a seedphrase.

JohnGuilding replied:
Yeah I agree. I think the best option is to find a way to allow dedicated parameters where there is no ambiguity over what the type should be. I think the solution we have now meets this goal, while allowing a user to create an account using either option.

blakecduncan reviewed (approved):
Nice, thanks for looking into the typescript stuff.  I added some comments to the PR.  Feel free to disregard my comments if you want to keep as is.  As I don't have a strong opinion.

JohnGuilding committed:
Simplify EOA account creation
- remove type brands
- remove casting
- use sensible defaults for top-level EOA account creation
- separate methods for account creation with private key or seed phrase
- add validate private key function
- add validate seed phrase function

JohnGuilding committed:
Move validate privateKey & seedPhrase functionality to dedicated functions

JohnGuilding commented on ethdk/.eslintrc.js:
```diff
@@ -16,5 +16,6 @@ module.exports = {
   rules: {
     indent: ['error', 2],
     'prettier/prettier': 'error',
+    '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'off',
```
Disabling this rule allows us to do this:
```ts
    if (validatedPrivateKey.success === false) {
      throw new Error(`Invalid private key: ${validatedPrivateKey.error}`)
    }
```

Without this rule, we get an error and the fix is this:
```ts
    if (!validatedPrivateKey.success) {
      throw new Error(`Invalid private key: ${validatedPrivateKey.error}`)
    }
```

I think the first option is more readable. This is fairly opinionated though so feel free to disagree with this change :)

blakecduncan replied:
I think that's a fair statement to say the first one is more readable.  I'm cool with disabling this rule 👍

blakecduncan reviewed (approved):
👍  lgtm!  Cheers for the updates
